### PR TITLE
RUN-3588 add support for registerUser API

### PR DESCRIPTION
**Pull request by patpao**
https://appoji.jira.com/browse/RUN-3588

This PR is for the new "registerUser" API that will replace the currently unused registerCustomData (see: https://github.com/openfin/Internal-Wiki/wiki/Custom-License-Registration)

Tests:
[W10](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/5a25cfba299179466481952a)
[W7](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/5a25cf3e2991794664819529)
[New Test](https://testing-dashboard.openfin.co/#/app/tests/5a25ce482991794664819528/edit) (I have more tests that need to be made public once the new RVM goes stable)

Related javascript-adapter PR: https://github.com/openfin/javascript-adapter/pull/365

### CODE REVIEW

File: src/browser/api/application.js [hunk 1]
```diff
@@ -371,28 +372,55 @@ Application.pingChildWindow = function() {
...
 
     if (!app) {
-        errorCallback(new Error(`application with uuid ${identity.uuid} does not exist`));
+        errorCallback(new Error(`application with uuid ${uuid} does not exist`));
+        return;
+    }
+
+    let licenseKey = app.licenseKey;
+    let configUrl = app._configUrl || app.configUrl;
+
```
Comment (wenjunche) [inline src/browser/api/application.js]: coreState.getConfigUrlByUuid(uuid)

Comment (patpao) [reply src/browser/api/application.js]: Thanks, fixed

File: src/browser/api/application.js [hunk 2]
```diff
@@ -371,28 +372,55 @@ Application.pingChildWindow = function() {
     console.warn('Deprecated');
 };
 Application.registerCustomData = function(identity, data, callback, errorCallback) {
-    let app = Application.wrap(identity.uuid);
+    console.warn('Deprecated');
```
Comment (HarsimranSingh) [inline src/browser/api/application.js]: Don't need this here. Instead put it into the javascript-adapter

File: src/browser/api/application.js [hunk 3]
```diff
@@ -371,28 +372,55 @@ Application.pingChildWindow = function() {
...
-            callback();
-        } else {
-            errorCallback(new Error('there was an issue sending a message to the RVM'));
+        if (!(uuid in registeredUsersByApp)) {
+            registeredUsersByApp[uuid] = new Set();
         }
+        registeredUsersByApp[uuid].add(userName);
+
+        sendToRVM({
+                topic: 'application',
```
Comment (HarsimranSingh) [inline src/browser/api/application.js]: Indentation in this block is a little wonky.

File: src/browser/api/application.js [hunk 4]
```diff
@@ -371,28 +372,55 @@ Application.pingChildWindow = function() {
     console.warn('Deprecated');
 };
 Application.registerCustomData = function(identity, data, callback, errorCallback) {
-    let app = Application.wrap(identity.uuid);
+    console.warn('Deprecated');
+    errorCallback(new Error('This function has been deprecated - please use Application.registerUser instead.'));
+};
+
+Application.registerUser = function(identity, userName, appName, callback, errorCallback) {
+    let uuid = identity.uuid;
```
Comment (HarsimranSingh) [inline src/browser/api/application.js]: const

File: src/browser/api/application.js [hunk 5]
```diff
@@ -371,28 +372,55 @@ Application.pingChildWindow = function() {
...
+    let app = coreState.getAppByUuid(uuid) || coreState.getExternalAppObjByUuid(uuid);
```
Comment (HarsimranSingh) [inline src/browser/api/application.js]: const

File: src/browser/api/application.js [hunk 6]
```diff
@@ -371,28 +372,55 @@ Application.pingChildWindow = function() {
...
 
     if (!app) {
-        errorCallback(new Error(`application with uuid ${identity.uuid} does not exist`));
+        errorCallback(new Error(`application with uuid ${uuid} does not exist`));
+        return;
+    }
+
+    let licenseKey = app.licenseKey;
```
Comment (HarsimranSingh) [inline src/browser/api/application.js]: const

File: src/browser/api/application.js [hunk 7]
```diff
@@ -371,28 +372,55 @@ Application.pingChildWindow = function() {
...
+    let configUrl = coreState.getConfigUrlByUuid(uuid);
```
Comment (HarsimranSingh) [inline src/browser/api/application.js]: const

File: src/browser/api/application.js (hunk 3 above)
Comment (patpao) [reply src/browser/api/application.js]: Every time I try to fix it, it seems like the "grunt deploy" command just undoes my changes

**Comment by patpao** [comment]
Updated - changed some variables to const in registerUser and removed registerCustomData completely (deprecated it in the javascript-adapter instead)

**Pull request merged**